src/DataLogger/DataLogger.cpp

```diff
@@ -0,0 +1,248 @@
+        if (columnsInFile == numColumns) {
+            // file is good
+            _numColumns = numColumns;
+            _curFilePtr = &tempFile;
+            return true;
+        }
+        // check if it is an empty file
+        else if (_numColumns == 0 && escaped == false)
+        {
+            // file is empty
+            _numColumns = numColumns;
+            _curFilePtr = &tempFile;
+            return true;
+        }
+        else 
+        {
+            // file is not empty, and does not have the correct number of columns
+            DEBUG_SERIAL_LN("Incorrect column number. Creating in file copyof_" + name);
+            tempFile.close();
+
+            // create new name for file
+
+            // find extension (example: myFile.csv)
+            int extensionIndex = name.lastIndexOf('.');
+            String base = name.substring(0, extensionIndex == -1 ? name.length() : extensionIndex);
+            String extension = extensionIndex == -1 ? "" : name.substring(extensionIndex + 1);
+            base.trim();
+
+            // find current file count (example: myFile (2).csv)
+            int fileCountStart = base.lastIndexOf('(');
+            int fileCountEnd = base.lastIndexOf(')');
+            int count = 0;
+            if (fileCountStart != -1 && fileCountEnd != -1 && fileCountEnd == base.length() - 1) {
+                String countString = base.substring(fileCountStart + 1, fileCountEnd);
+                bool numeric = true;
+                
+                // checks if all characters in countString are numeric
+                for (size_t i = 0; i < countString.length(); i++) {
+                    if (!isdigit(countString.charAt(i))) {
+                        numeric = false;
+                        break;
+                    }
+                }
+
+                if (numeric) {
+                    count = countString.toInt();
+                    // take count out of file name
+                    base = base.substring(0, fileCountStart);
+                    base.trim();
+                }
+            }
+            String newFileName = base + " (" + (count + 1) + ")" + (extension != "" ? "." + extension : "");
+            return open(newFileName, numColumns);
+        }
+    }
+    else
+    {
+        // file doesnt exist. create new file
+
+        DEBUG_SERIAL_LN(name + " doesn't exist. Creating new file");
+        return create(name, numColumns);
+    }
+}
+
+bool DataLogger::saveToDisk() {
+    if (_curFilePtr != nullptr)
+    {
+        (*_curFilePtr).print(_buffer);
+        _buffer = "";
+        DEBUG_SERIAL_LN("Write to file successful.");
+        return true;
+    }
+    else
+    {
+        DEBUG_SERIAL_LN("Write to file failed.");
+        return false;
+    }
+}
+
+bool DataLogger::close()
+{
+    if (_curFilePtr != nullptr)
+    {
+        (*_curFilePtr).close();
+
+        // sets _curFile to null so saveToDisk() doesnt try writing to a closed file
+        _curFilePtr = nullptr;
```

Comment:
I changed ```_curFile``` to ```_curFilePtr``` because objects cannot be set to null, while pointers can.

@cosparks do we have to run something like ```delete (*_curFilePtr)``` to prevent memory leaks?

Reply:
You only need to use the delete operator if you've used the new operator somewhere.  In general, you only need to worry about memory leaks if you're calling new or any of the C dynamic memory allocation functions (malloc, alloc etc..)

Reply:
Ah okay thats helpful, thanks